[RNMobile] Update Image component to avoid flicker when updating the URI

derekblank opened the pull request:
## What?
Fixes an issue where image uploads disappear or "blink" when transitioning from local to network states.

Related PRs:
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6550
* https://github.com/wordpress-mobile/WordPress-Android/pull/19963
* https://github.com/wordpress-mobile/WordPress-iOS/pull/22409

## Why?
Fixes:
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/6510
* https://github.com/wordpress-mobile/WordPress-iOS/issues/21358

Currently when uploading an Image, the Image itself "blinks" when transitioning its URI reference from a local on-device image to the uploaded, network-available image. This "blink" is possibly due to the time spent downloading the network-available resource itself, where the visual asset appears unavailable while the network image is downloading.

## How?
Adds more descriptive states when a `uri` is changed to await the full loading of a network-based image, before swapping out the URI from the local filestore-based image.

## Testing Instructions
1. Create a post and add an Image or Gallery block
2. Select "Choose from Device" from the media upload options and upload an image
3. When the image has been uploaded, observe that it does not blink or disappear when transitioning to full opacity

## Screenshots or screencast <!-- if applicable -->
Before | After
-|-
<img alt="A gif of an image blinking" src="https://github.com/WordPress/gutenberg/assets/643285/c77768d1-6e46-41b0-8ebe-64f74104bdb5" /> | <img alt="A gif of an image with a smooth transition" src="https://github.com/WordPress/gutenberg/assets/643285/d265fcca-63c1-4267-8956-f2013a4defd4" />

derekblank committed:
Update native image component to avoid flicker when updating the URI

derekblank committed:
Update non-visible Image block styles

derekblank committed:
Update mobile Image block to use separate network and local image URL behavior between platforms

derekblank committed:
Update Image block platform fetching logic

derekblank committed:
Update Android logic for Image block platform fetching logic

derekblank committed:
Add further updates to Image block platform logic

derekblank committed:
Fix code formatting issue in Image block source prop

derekblank committed:
Resolve RNImage.prefetch promise

derekblank committed:
Update CHANGELOG

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -80,6 +83,22 @@ const ImageComponent = ( {
 				}
 			} );
+
+			if ( url.startsWith( 'file:///' ) ) {
+				setLocalURL( url );
+			} else if ( url.startsWith( 'https://' ) ) {
+				if ( Platform.isIOS ) {
+					setNetworkURL( url );
+				} else if ( Platform.isAndroid ) {
+					const result = RNImage.prefetch( url );
+					if ( result ) {
+						RNImage.prefetch( url ).then( () => {
+							setNetworkURL( url );
+							setNetworkImageLoaded( true );
+						} );
+					}
+				}
+			}
```
I feel we should separate this logic from this existing effect, as it appears unrelated to the existing effect. I.e., relocate it to a separate `useEffect`. 

It might also be worth considering abstracting it into a custom Hook. Benefits from doing so might be communicating its intent with a name, abstracting complexity, and simplifying automated unit testing. I am picturing something like the following, where all the state and effects are managed within the custom Hook. 

```js
const { error, isLoading, url } = usePrefetchedUrls(url);
```

WDYT?

derekblank replied:
> I feel we should separate this logic from this existing effect, as it appears unrelated to the existing effect. I.e., relocate it to a separate useEffect.

The two parts of the useEffect are somewhat related as they are coordinating the layout of the Images based on the URL. The sole dependency of the useEffect is `url`, and the output of the original part of the useEffect was to set the `imageData` value, which is used to transition between the image placeholder and the image components themselves:

```
{ ! imageData ? (
	// Static placeholder icon)
) : (
	// image components where `imageStyles` is derived from `imageData` state
        // <Image style={ imageStyles } />
) }
```

However, I follow about the separation of concerns between prefetching URLs and setting image metadata, and I also agree with the clarity of using a hook's name to communicate intent of the logic.

I'll try updating to a hook to see the results. Would you expect this hook to live outside this file? I don't think I would, as the logic seems isolated to this component, but possibly not.

dcalhoun replied:
> The two parts of the useEffect are somewhat related as they are coordinating the layout of the Images based on the URL.

After considering the examples you provided, I agree the existing `imageData` fetch and URL prefetch do relate to one another. 

We could (A) treat them as separate effects and compose their results together in the render or (B) combine them into a single effect and compose their result as the Hook return value. It is not clear to me whether one approach is more appropriate than the other. I open to either approach. 

If we combined them into a single Hook, I suppose it might resemble something like the following. 

```js
const { error, isLoading, data } = usePrefetchedImage(url);
// console.log(data); returns...
// {
//   src: "https://example.com/image.jpg",
//   width: 100,
//   height: 100,
//   aspectRatio: 1,
// }
```

> In order to retrieve the image dimensions, the image may first need to be loaded or downloaded, after which it will be cached. This means that in principle you could use this method to preload images, however it is not optimized for that purpose, and may in future be implemented in a way that does not fully load/download the image data. A proper, supported way to preload images will be provided as a separate API.

I will also note that the [`getSize` method notes](https://reactnative.dev/docs/image#getsize) the above, which could mean that we do not need to call both `getSize` and `prefetch`...I'm not sure. 

> Would you expect this hook to live outside this file?

No, I envisioned placing it within the same file outside of the component function.

derekblank replied:
> We could (A) treat them as separate effects and compose their results together in the render or (B) combine them into a single effect and compose their result as the Hook return value.

Given that there isn't a major technical advantage between these two options, I think option (A) "treat them as separate effects and compose their results together in the render" would be the most straightforward. The separation of concerns feel separate enough that they could be easily understood with improving the code commenting here, which I'll do. 

It's also worth noting that we can consider this solution a temporary improvement for the user experience while we explore a more sustainable technical solution for replacing FastImage and realigning the Platform divergences. My hope is that the logic introduced in this PR will not be necessary long-term with an improved image component.

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -80,6 +83,22 @@ const ImageComponent = ( {
+					const result = RNImage.prefetch( url );
+					if ( result ) {
+						RNImage.prefetch( url ).then( () => {
+							setNetworkURL( url );
+							setNetworkImageLoaded( true );
```
We may should guard against setting local state for "unmounted" components after the asynchronous prefetch. I.e., utilize an `isCurrent` guard like the `imageMetaData` effect.

derekblank replied:
Do you mean utilizing `isCurrent` again like it is [here](https://github.com/WordPress/gutenberg/blob/7b2350ae84d73957072282ae388aefd35f885825/packages/components/src/mobile/image/index.native.js#L70-L72)?

I.e., something like:

```
if ( url.startsWith( 'file:///' ) ) {
	setLocalURL( url );
} else if ( url.startsWith( 'https://' ) ) {
	if ( Platform.isIOS ) {
		setNetworkURL( url );
	} else if ( Platform.isAndroid ) {
		const result = RNImage.prefetch( url );
		if ( result ) {
			if ( ! isCurrent ) { // <!--------------------- This 
				return;
			}
			RNImage.prefetch( url ).then( () => {
				setNetworkURL( url );
				setNetworkImageLoaded( true );
			} );
		}
	}
}
```

dcalhoun replied:
Correct, I meant leveraging an `isCurrent` value to guard against setting local state for (1) a component that is no longer mounted after the asynchronous fetch completes or (2) an effect that was replaced by a new effect invocation. 

Your example is close to what I pictured, but it caused me to note there are currently two invocations of `RNImage.prefetch` in the proposed changes. Why is that? Presumably we only need one invocation. 

```js
useEffect(() => {
  let isCurrent = true;

  if (Platform.isIOS) {
    // iOS logic
  } else if (Platform.isAndroid) {
    RNImage.prefetch(url).then(
      () => {
        if (!isCurrent) {
          return;
        }
        setNetworkURL(url);
        setNetworkImageLoaded(true);
      },
      (error) => {
        // handle error
      }
    );
  }

  return () => {
    isCurrent = false;
  };
}, [url]);
```

derekblank replied:
Good catch. `const result = RNImage.prefetch( url );` was added to resolve unit test failures with resolving Promises correctly. The duplicative `RNImage.prefetch( url )` statements are not necessary, and using `(error) => { }` is probably a better handler. 

I'll update, but in trying to consider what we would throw as an error from `(error) => { }`, my thought is perhaps we throw nothing when the networkURL prefetch fails, and just set the localURL state. My reasoning is this code is focused on displaying the response of the image upload in the UI, but is not meant to handle the failure of the upload itself -- in the event of a true network failure, the `EDITOR_UPLOAD_MEDIA_FAILED` event would be invoked, and the local image would persist in the UI as `localURL` until the upload was either retried or dismissed.

dcalhoun replied:
> My reasoning [for not communicating the error to the users] is this code is focused on displaying the response of the image upload in the UI, but is not meant to handle the failure of the upload itself [...]

I follow your rationale. I will note some nuance in the relationship between this code and the upload. My understanding is that this logic will only execute when a remote `url` is provided to the component, i.e., the upload successfully completed. 

This scenario where the upload succeeds and the prefetch fails means the file exists on the server but the UI was unable to fetch it to display it. We likely have a few options: 

1.  Retain the local URL in state, displaying the local cache until the remote is successfully fetched (on next post open?). 
2. Retain the local URL and retry in the background (on some interval, to some maximum attempts). 
3. Retain the local URL and display a notice with a button to retry. 

The second option feels like somewhat unnecessary complexity. The third feels like it might cause unnecessary confusion for the user. The first may be most suitable, but I wonder if there are scenarios where this odd state breaks other functionality, e.g., editing the image. 

I recommend simulating this scenario and exploring the app features and how the experience feels. 

WDYT?

derekblank replied:
I went through quite a few permutations of the ideas above -- from the results, I don't think we're going to find a perfect solution. Until we find a full fix for swapping out the `source` prop in our image components, I think this will always represent a tradeoff. The changes introduced in this PR reduce the flicker effect in most circumstances.

I am still able to reproduce a flicker under certain circumstances when replacing a media upload that has already been uploaded with the same image file. This matches the current behavior, but could possibly be improved through further debugging of `setNetworkUrl` state.

derekblank replied:
I noted that using the syntax suggested in [this comment](https://github.com/WordPress/gutenberg/pull/57869#discussion_r1462185577) will produce a failure when running the Gallery block tests:

```
useEffect(() => {
  let isCurrent = true;

  if (Platform.isIOS) {
    // iOS logic
  } else if (Platform.isAndroid) {
    RNImage.prefetch(url).then(
      () => {
        if (!isCurrent) {
          return;
        }
        setNetworkURL(url);
        setNetworkImageLoaded(true);
      },
      (error) => {
        // handle error
      }
    );
  }

  return () => {
    isCurrent = false;
  };
}, [url]);
```

derekblank replied:
@dcalhoun In reference to the above notion:

>  I don't think we're going to find a perfect solution

At the moment, I feel I'm at an impasse in moving this work forward. Do you feel there is value in merging/continuing any of this work? In my testing, this work reduces image flicker overall for new file uploads, but there appears to be edge/caching edge cases for when the same file is repeatedly uploaded and replaced, which may be acceptable. But, considering that this work should eventually be replaced by a more performant image component, I'm questioning if this path identified in this PR is worth pursuing. 

At the moment, I am optimistic that this change would provide value for the user when uploading media as a short-term fix, but am also trying to identify any potential regressions from the current behavior. So far, I haven't noted any regressions that are not already present in media upload flows. If you do not note any regressions, perhaps we could lean into the [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement) aspect of this fix.

dcalhoun replied:
> I noted that using the syntax suggested in [this comment](https://github.com/WordPress/gutenberg/pull/57869#discussion_r1462185577) will produce a failure when running the Gallery block tests:

I opened https://github.com/WordPress/gutenberg/pull/58213 addressing the test failures for your consideration. 

> The changes introduced in this PR reduce the flicker effect in most circumstances.

> I am still able to reproduce a flicker under certain circumstances when replacing a media upload that has already been uploaded with the same image file. This matches the current behavior, but could possibly be improved through further debugging of `setNetworkUrl` state.

> At the moment, I am optimistic that this change would provide value for the user when uploading media as a short-term fix, but am also trying to identify any potential regressions from the current behavior.

If the proposed changes improve (what is likely) the most used experience (inserting new photos) _without introducing regressions_, then I think it is worth merging this incremental improvement. I agree that it is not the long-term solution, but we do not know how soon we will implement a more robust solution. I truly do believe you are proposing a better UX from what I saw in testing. 👏🏻 

I am happy to help regression test if you are willing to create new prototype builds with the implementation you consider the "final" approach.

derekblank replied:
Thanks for sharing your perspective and for fixing the unit test approaches. I will update the prototype builds and run through our media-focused [test cases](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases) on a non-debug build, and post the results. Any extra regression testing would be welcome. 👍

derekblank replied:
I went through all of the image test cases to check for regressions and noted my results:

- [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/22409#issuecomment-1909337182)
- [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android/pull/19963#issuecomment-1909293302)

I did note a minor regression with the Media & Text and Cover blocks where the image was uploaded successfully but the image was not displayed in the Editor unless the post was closed and re-opened. I traced this to the `onLoad` event not firing from the nonVisibleImage to set `networkImageLoaded` to `true`. When the styles were updated to use `1` for the width and height of the nonVisibleImage, the `onLoad` event fired and everything worked correctly. So, apparently the workaround that applied to Android in [this comment](https://github.com/WordPress/gutenberg/pull/57869#discussion_r1462849227) about the image needing non-zero width/height values to fire `onLoad` also applies to iOS too. 

Adding the 1px width/height value causes the blue outline to be offset by 1px as well when the block is selected, which we'll need to fix. Other than that, I am feeling much more confident in the feature after performing the regression testing.

derekblank replied:
@dcalhoun I have updated the [iOS build](https://github.com/wordpress-mobile/WordPress-iOS/pull/22409#issuecomment-1897688547) after fixing the Media-Text and Cover block display regression. If you have a vision for a potential hook solution and would like to make an update, please feel free. If not, we can make a final regression testing and move forward with this solution as an incremental update.

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -259,14 +289,57 @@ const ImageComponent = ( {
+										uri:
+											networkURL && networkImageLoaded
+												? networkURL
+												: ( localURL && localURL ) ||
+												  url,
```
Can this be simplified as `localUrl || url`? The duplicate references to `localURL` seems unnecessary.

derekblank replied:
Yes, I agree. Will test and update. 👍

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -259,14 +289,57 @@ const ImageComponent = ( {
+									source={ { uri: networkURL } }
+									style={ styles.nonVisibleImage }
+									onLoad={ () => {
+										setNetworkImageLoaded( true );
```
I was trying to consider if `networkImageLoaded` could be removed entirely as it seems a bit redundant, but I suppose it is necessary for the iOS implementation specifically. Correct?

derekblank replied:
Correct. On iOS, `networkURL` alone cannot be used as a boolean to trigger that the network image is ready to be displayed. There is a period of time where `networkURL` is known to the component as a string that _could_ be used as a boolean (`https://site.com/network-image.jpg`), but the image data itself has not been downloaded. In the current iOS implementation, the brief period of time where this occurs is the source of the disappearing image.

In essence, the lack of a callback from the FastImage [`preload`](https://github.com/DylanVann/react-native-fast-image/?tab=readme-ov-file#fastimagepreload-source--void) method is the root of the entire problem on iOS, and the reason for the divergence between the two platforms. My perception is that FastImage's preload method was simply not designed for our use case of setting local image data and replacing it with network image data -- rather, it seems designed for use with static placeholders to be replaced by network image data as soon as it's available.

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -259,14 +289,57 @@ const ImageComponent = ( {
-						/>
+						{ Platform.isAndroid && (
+							<>
+								{ networkImageLoaded && networkURL && (
+									<Animated.Image
+										style={ imageStyles }
+										fadeDuration={ 0 }
+										source={ { uri: networkURL } }
+										{ ...( ! focalPoint && {
+											resizeMethod: 'scale',
+										} ) }
+										resizeMode={ imageResizeMode }
+									/>
+								) }
+								{ ! networkImageLoaded && ! networkURL && (
+									<Animated.Image
+										style={ imageStyles }
+										fadeDuration={ 0 }
+										source={ { uri: localURL } }
+										{ ...( ! focalPoint && {
+											resizeMethod: 'scale',
+										} ) }
+										resizeMode={ imageResizeMode }
+									/>
+								) }
```
Are the sibling `Animated.Image` components necessary for transition purposes or could we achieve the same animation by flipping `source` on a single `Animation.Image`?

derekblank replied:
The flickering still occurs when only one RNImage instance is used and the `source` prop is changed:

```
<Animated.Image
    source={ { uri: networkURL ? networkURL : localURL } }
/>
```

My understanding is that the flickering will always occur on Android when the `source` prop is changed. While it's very noticeable when using RNImage, it also affects FastImage on Android as well (as described in https://github.com/DylanVann/react-native-fast-image/issues/747). It appears to be related to something more fundamental on how image data is cached and retrieved on Android. The apparent solution is to use separate image instances to avoid changing the `source` prop entirely.

dcalhoun reviewed (commented):
This turned out to be a complex subject. 😅 From my initial testing, the functionality works as described. Well done! 👏🏻 

I did notice that replacing media for an Image or Gallery block does not show the _newly selected_ local image immediately. Instead, it displays the previous image until the upload completes, it then swaps to the new image instantly. I'm not sure if this an existing issue or not. 

I also noticed an image can flash when the upload finishes in a Gallery block, but only on occasion. It occurs sometimes when I upload ~4 images to a new Gallery block. It could be unrelated to the Gallery block and I just happen to recreate it more frequently with Gallery block given it generally houses multiple images.

derekblank committed:
Update non-visible image styles

Resolves layout display issue when block is selected

derekblank commented:
> I did notice that replacing media for an Image or Gallery block does not show the newly selected local image immediately. Instead, it displays the previous image until the upload completes, it then swaps to the new image instantly. I'm not sure if this an existing issue or not.

@dcalhoun Good catch, I also noted that the newly selected local image does not immediately replace the local file until it's uploaded to the server. While it still prevents the flickering, I think it's worth seeing if we can improve the replace behavior as well.

dcalhoun commented on packages/components/src/mobile/image/style.native.scss:
```diff
@@ -171,3 +171,9 @@
 }
+
+.nonVisibleImage {
+	height: 0;
+	width: 0;
```
> height and width must be non-zero or else onLoad does not fire on Android 

My understand is that using `0` for these values is acceptable because the above inline comment sourced from the [workaround](https://github.com/DylanVann/react-native-fast-image/issues/747#issuecomment-1264047430) does not apply as we do _not_ use FastImage for Android. Correct?

derekblank replied:
This is correct. I modified these to use `0`, as using `1` impacted how the blue border was displayed around the image when the block was selected, and was also unnecessary from a functionality perspective for the reasons mentioned in the [workaround](https://github.com/DylanVann/react-native-fast-image/issues/747#issuecomment-1264047430) that only apply to Android.

derekblank replied:
For posterity and anyone finding this thread in the future, it turns out that a value of at least `1` must be used for both iOS and Android in order for the `onLoad` event to fire from FastImage.

derekblank committed:
Remove duplicate references to localURL

derekblank committed:
Fix Replace Image behavior to update from localURL

derekblank committed:
Refactor Image block URL handling logic

dcalhoun committed:
test: Fix failures due to image component modifications (#58213)

The mobile image component introduce new asynchronous effects. We must
await the result of those effect, in the form of asserting UI changes
from the subsequent state updates, to satisfy the React test utilities
expectation that unexpected re-renders do not occur after the test
completes.

Additionally, there were instances of awaiting `act` invocations that
were not asynchronous. The erroneous usage of `await` for these
synchronous `act` calls resulted in cryptic errors. In reality, the
logic run within these `act` calls are synchronous and should merely be
wrapped in `act` to signal that they result in a re-render.

derekblank committed:
Update nonVisibleImage styles

Using width: 1 and height: 1 ensures that onLoad will fire

derekblank committed:
Merge branch 'rnmobile/prevent-image-flicker-when-uploading' of https://github.com/wordpress/gutenberg into rnmobile/prevent-image-flicker-when-uploading

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -80,6 +83,27 @@ const ImageComponent = ( {
+							setNetworkImageLoaded( true );
+						},
+						() => {
+							// handle error
```
If we do not intend to handle this error, we should update this inline comment stating why the error is not handled or remove the no-op callback.

derekblank replied:
Updated with a code comment that this UI state is handled by `isUploadFailed` event. Ignoring the error here will persist the local URI in the UI.

dcalhoun commented on packages/components/src/mobile/image/index.native.js:
```diff
@@ -80,6 +83,27 @@ const ImageComponent = ( {
 			} );
+
+			if ( url.startsWith( 'file:///' ) ) {
+				setLocalURL( url );
```
My guess is that this replacing images does not immediate display the new local image or perform smooth transitions to the network image because the `networkUrl` and `networkImageLoaded` values are not reset when a new, replacement local URL is provided. I wonder if we could resolve those issues by reseting the values here. 

From my initial testing, the following improves the situation. WDYT? 

```suggestion
				setLocalURL( url );
				setNetworkURL( null );
				setNetworkImageLoaded( false );
```

derekblank replied:
Good suggestion. I had explored this idea initially, but my approach had some unintended effects on Android. This appears to mitigate the issue for both platforms, however, so I'll commit it and do further testing for both. 🙇

dcalhoun replied:
> I had explored this idea initially, but my approach had some unintended effects on Android. This appears to mitigate the issue for both platforms, however, so I'll commit it and do further testing for both. 🙇

🤔 Alright, please be sure to retest these changes to verify it does not introduce the unintended effect on Android. I am unfamiliar with what specifically that may be.

dcalhoun reviewed (commented):
I did not note any new regressions in my testing. During my testing I noted:

* [Known Regression] Replacing an image does not display the new image until the upload is completed, which is a regression that we may choose to ship. I left an inline comment regarding a potential solution. 
* On iOS only, Media + Text and Cover blocks do not display the local image during the upload, but this appears to be true in the production app as well. 
* Choosing a photo from a "free photo library" does not result in a smooth transition.

derekblank committed:
Update packages/components/src/mobile/image/index.native.js

Co-authored-by: David Calhoun <github@davidcalhoun.me>

derekblank committed:
Update RNImage error handler event with code comment

derekblank committed:
Merge branch 'rnmobile/prevent-image-flicker-when-uploading' of https://github.com/wordpress/gutenberg into rnmobile/prevent-image-flicker-when-uploading

dcalhoun reviewed (approved):
The proposed changes look good to me. 

I have not retested with the latest commits after https://github.com/WordPress/gutenberg/pull/57869#pullrequestreview-1844370772, as there is not an updated prototype build. However, if you'd like for me to re-test something, please let me know. 

I will also note that we should re-test for any Android regressions you noted in https://github.com/WordPress/gutenberg/pull/57869#discussion_r1471177080. 

Thank you for improving this image upload experience!

derekblank committed:
Update selected images styles on iOS to account for non-visible image

On iOS, add 1 to height to account for the 1px non-visible image
that is used to determine when the network image has loaded
We also must verify that it is not NaN, as it can be NaN when the image is loading.

derekblank commented:
> I will also note that we should re-test for any Android regressions you noted in https://github.com/WordPress/gutenberg/pull/57869#discussion_r1471177080.

I ran through the regression tests for both platforms, and did not note any with the updated logic. Your logic change also improved the replacement behavior for Android -- thanks for the contribution!

I did note that when an image is selected on iOS, there is a 1px offset between the border and the image due to the non-visible image. I noted this as something to fix in a previous comment, and updated the `imageSelectedStyles` to account or this on iOS. 

<details>
<summary>Before | After</summary>

Note the 1px offset on the bottom of the selected image in the "Before" example:

Before | After
-|-
<img src="https://github.com/WordPress/gutenberg/assets/643285/4f82b668-52eb-4fe3-9271-c09206dd36cb" /> | <img src="https://github.com/WordPress/gutenberg/assets/643285/2fed8c65-7847-4d32-93c1-361e0a845fa4" />

</details>

derekblank commented:
@dcalhoun I have started the main apps build CI jobs to reference the latest changes, and tested both platforms successfully locally. Although you've already approved, I just wanted to provide a final chance for any review from the last change to update the selected image styles, and general testing using the builds (which should have succeeded by the time you read this).

Tomorrow I intend to test the main apps CI builds and if successful, merge the Gutenberg / Gutenberg Mobile PRs before cutting the 1.112.0 release. Thanks for your feedback and contributions to this fix. 🙇 

- [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android/pull/19963#issuecomment-1895651530)
- [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/22409#issuecomment-1897688547)

dcalhoun reviewed (approved):
Latest changes look good and tested well for me on iOS and Android. Hopefully we can refactor/remove this complexity in the future. Thanks for seeing this improvement through!
